extensions/amp-consent/0.1/amp-consent.js

```diff
@@ -18,34 +18,203 @@ import {CSS} from '../../../build/amp-consent-0.1.css';
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_POLICY_MANGER)
+        .then(manager => {
+          this.consentPolicyManager_ = manager;
+          this.generateDefaultPolicy_();
+          const policyKeys = Object.keys(this.policyConfig_);
+          for (let i = 0; i < policyKeys.length; i++) {
+            this.consentPolicyManager_.registerConsentPolicyInstance(
+                policyKeys[i], this.policyConfig_[policyKeys[i]]);
+          }
+        });
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_STATE_MANAGER)
+        .then(manager => {
+          this.consentStateManager_ = manager;
+          this.init_();
+        });
+
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout == Layout.NODISPLAY;
   }
+
+  /**
+   * Init the amp-consent by registering and initiate consent instance.
+   */
+  init_() {
+    this.cidPromise_ = Services.cidForDoc(this.getAmpDoc()).then(cid => {
+      // Note: do not wait for consent of the cid used for amp-consent
+      return cid.get({
+        scope: TAG,
+        createCookieIfNotPresent: true,
+      }, Promise.resolve());
+    });
+
+    const instanceKeys = Object.keys(this.consentConfig_);
+    for (let i = 0; i < instanceKeys.length; i++) {
+      const instanceId = instanceKeys[i];
+      this.consentStateManager_.registerConsentInstance(instanceId);
+      this.getConsentRemote_(instanceId).then(response => {
+        this.parseConsentResponse_(instanceId, response);
+        this.handleUI_(instanceId);
+      }).catch(unusedError => {
+        // TODO: Handle errors
+      });
+    }
+  }
+
+  /**
+   * Generate default consent policy if not defined
+   */
+  generateDefaultPolicy_() {
+    if (this.policyConfig_ && this.policyConfig_['default']) {
+      return;
+    }
+    // Generate default policy
+    const instanceKeys = Object.keys(this.consentConfig_);
+    const defaultWaitForItems = {};
+    for (let i = 0; i < instanceKeys.length; i++) {
+      // TODO: Need to support an array.
+      defaultWaitForItems[instanceKeys[i]] = undefined;
+    }
+    const defaultPolicy = {
+      'itemsToWait': defaultWaitForItems,
+    };
+    this.policyConfig_['default'] = defaultPolicy;
+  }
+
+  /**
+   * Get localStored consent info, and send request to get consent from endpoint
+   * @param {string} instanceId
+   * @return {!Promise<!JsonObject>}
+   */
+  getConsentRemote_(instanceId) {
+    return this.consentStateManager_.getConsentInstanceState(instanceId)
+        .then(localState => {
+          return this.cidPromise_.then(ampUserId => {
+            const request = /** @type {!JsonObject} */ ({
+              'ampUserId': ampUserId,
+              'consentInstanceId': instanceId,
+              'consentState': localState,
+            });
+            // TODO: Decide which request method/Content Type to use.
+            const init = {
+              headers: {
+                'Content-Type': 'application/json',
```

Comment:
would this cause a preflight? see comment in xhr-impl.js L447

Reply:
If you change it to `text/html` you may need to add another [body parser middleware](https://github.com/expressjs/body-parser#bodyparsertextoptions)

Reply:
Interesting I didn't see a preflight. But I did change to use `test/plain`. 
True I need to add another body parser middleware, but I am not reading the request body in this PR. I can add the middleware in following PR : )